torch/quantization/_correct_bias.py

```diff
@@ -0,0 +1,292 @@
+from __future__ import print_function, division, absolute_import
+import torch
+import torch.nn as nn
+import torch.nn.quantized as nnq
+
+import torch.quantization
+import torch.quantization._numeric_suite as ns
+
+import copy
+
+_supported_modules = {nn.Linear, nnq.Linear, nn.Conv2d, nnq.Conv2d}
+
+def get_module(model, name):
+    ''' Given name of submodule, this function grabs the submodule from given model
+    '''
+    curr = model
+    name = name.split('.')
+    for subname in name:
+        if subname == '':
+            return curr
+        curr = curr._modules[subname]
+    return curr
+
+def get_parent_module(model, name):
+    ''' Given name of submodule, this function grabs the parent of the submodule, from given model
+    '''
+    curr = model
+    name = name.split('.')[:-1]
+    for subname in name:
+        if subname == '':
+            return curr
+        curr = curr._modules[subname]
+    return curr
+
+def get_param(module, attr):
+    ''' Sometimes the weights/bias attribute gives you the raw tensor, but sometimes
+    gives a function that will give you the raw tensor, this function takes care of that logic
+    '''
+    param = getattr(module, attr, None)
+    if isinstance(param, nn.Parameter):
+        return param
+    elif param is not None:
+        return param()
+
+
+class MeanLogger(ns.Logger):
+    def __init__(self):
+        super(MeanLogger, self).__init__()
+        self.stats["tensor_val"] = None
+        self.count = 0
+        self.sum = None
+
+    def forward(self, x):
+        ''' takes the rolling mean of the data being passed through
+        '''
+        y = copy.deepcopy(x)
+        if x.is_quantized:
+            x = x.dequantize()
+        if self.stats["tensor_val"] is None:
+            self.stats["tensor_val"] = x
+            self.sum = x
+            self.count = 1
+        else:
+            torch.Tensor.add_(self.sum, x)
+            self.count += 1
+            self.stats["tensor_val"] = self.sum / self.count
+        return y
+
+class ShadowModule(nn.Module):
+    def __init__(self, float_module=None, quantized_module=None):
+        super(ShadowModule, self).__init__()
+        self.float_module = float_module
+        self.quantized_module = quantized_module
+
+    def forward(self, x):
+        ''' Reads in a batch of data, using the computed means of the input and output of the
+        float and quantized modules, bias correction will be applied on the quantized module
+        '''
+        self.float_module(x.dequantize())
+        a = self.quantized_module(x)
+
+        output_logger, input_logger = get_inputs_n_outputs(self.float_module, self.quantized_module)
+        correct_quantized_bias(output_logger, input_logger, self.float_module, self.quantized_module)
+        return a
+
+def add_shadow(float_model, quantized_model, white_list=_supported_modules):
```

Comment:
why can't we use numeric suite to add shadow module?

Reply:
Yes, you should be calling: https://github.com/pytorch/pytorch/blob/master/torch/quantization/_numeric_suite.py#L307